<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!--NewPage-->
<HTML>
<HEAD>
<!-- Generated by javadoc (build 1.5.0_13) on Sat Sep 13 08:11:45 PDT 2008 -->
<TITLE>
SimilarityPair
</TITLE>

<META NAME="keywords" CONTENT="org.obd.model.stats.SimilarityPair class">

<LINK REL ="stylesheet" TYPE="text/css" HREF="../../../../stylesheet.css" TITLE="Style">

<SCRIPT type="text/javascript">
function windowTitle()
{
    parent.document.title="SimilarityPair";
}
</SCRIPT>
<NOSCRIPT>
</NOSCRIPT>

</HEAD>

<BODY BGCOLOR="white" onload="windowTitle();">


<!-- ========= START OF TOP NAVBAR ======= -->
<A NAME="navbar_top"><!-- --></A>
<A HREF="#skip-navbar_top" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_top_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="class-use/SimilarityPair.html"><FONT CLASS="NavBarFont1"><B>Use</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../../../../org/obd/model/stats/ScoredNodePair.html" title="class in org.obd.model.stats"><B>PREV CLASS</B></A>&nbsp;
&nbsp;<A HREF="../../../../org/obd/model/stats/SummaryStatistics.html" title="class in org.obd.model.stats"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../../index.html?org/obd/model/stats/SimilarityPair.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="SimilarityPair.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;FIELD&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;FIELD&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_top"></A>
<!-- ========= END OF TOP NAVBAR ========= -->

<HR>
<!-- ======== START OF CLASS DATA ======== -->
<H2>
<FONT SIZE="-1">
org.obd.model.stats</FONT>
<BR>
Class SimilarityPair</H2>
<PRE>
java.lang.Object
  <IMG SRC="../../../../resources/inherit.gif" ALT="extended by "><B>org.obd.model.stats.SimilarityPair</B>
</PRE>
<HR>
<DL>
<DT><PRE>public class <B>SimilarityPair</B><DT>extends java.lang.Object</DL>
</PRE>

<P>
This object represents a pair of comparable nodes (such as classes in ontologies) that can have other nodes in common (e.g. via annotation)
 
 For example:
 
 <ul>
 <li>two gene IDs that potentially shared ontology classes via annotations (e.g. two homologous genes in different species)
 <li>two annotation sources that annotate the same entity (e.g. in NCBO OMIM 3x annotation coverage experiment)
 <li>two classes in an ontology that potentially share ancestors
 </ul>
 
 <h2>Getting Started</h2>
 
 Use of this class assumes you already have two nodes you wish to compare. These might be two genotypes from the same gene - or two orthologous genes.
 This class will NOT perform search. For that you need  <A HREF="../../../../org/obd/query/AnalysisCapableRepository.html#getSimilarNodes(java.lang.String)"><CODE>AnalysisCapableRepository.getSimilarNodes(String)</CODE></A>. This class assumes you have the pair of nodes
 in advance.
 
 <h2>Basic Concepts</h2>
 
 Nodes are deemed similar on the basis of <i>what they have in common</i>. This class is flexible with respect to what kinds of properties can be used as
 a basis for similarity. However, the most common case is where the two nodes are entities that are annotated (such as genes), and we are looking for
 similarity on the basis of <i>shared annotations</i> to classes in an ontology, or to compositional description classes (see <CODE>CompositionalDescription</CODE>).
 <p>
 In these cases, we used <i>inferred</i> annotations. E.g. if geneA is annotated to Leg and geneB to Wing, they have Appendage in common.
 
 <h2>Node Sets</h2>
 
 Scoring is typically a measure of what the nodes have in common vs what one node has that the other one does not.
 <p>
 the set of classes annotated for node1 is referred to here as nodesInSet1 (accessed via <A HREF="../../../../org/obd/model/stats/SimilarityPair.html#getNodesInSet1()"><CODE>getNodesInSet1()</CODE></A>), and likewise for node2 the set is
 nodesInSet2 (accessed via <A HREF="../../../../org/obd/model/stats/SimilarityPair.html#getNodesInSet2()"><CODE>getNodesInSet2()</CODE></A>)
 <p>
 the set of classes in common to both node1 and node2 inferred annotations is called nodesInCommon (see <A HREF="../../../../org/obd/model/stats/SimilarityPair.html#getNodesInCommon()"><CODE>getNodesInCommon()</CODE></A>).
 We write this as:  
 <pre>
    nodesInSet1 &cap; nodesInSet2
 </pre>
 
 The size of this set is written as:
 
 <pre>
   | nodesInSet1 &cap; nodesInSet2 |
 </pre>
 
 The entire set of classes used to annotate both nodes (including inferred annotations) is called nodesInUnion (see <A HREF="../../../../org/obd/model/stats/SimilarityPair.html#getNodesInUnion()"><CODE>getNodesInUnion()</CODE></A>).
 We write this as:  
 
 <pre>
    nodesInSet1 &cup; nodesInSet2
 </pre>
 
 We size of this set is written as:
 
 <pre>
   | nodesInSet1 &cup; nodesInSet2 |
 </pre>
 
 Note that the size of this set is not the sum | nodesInSet1 | + | nodesInSet2 |, because there are usually some nodes in common
 <p>
 For the purposes of reporting similar annotations, we want to avoid reporting redundant classes; e.g. two genes have annotations to "femur" they also
 have inferred annotations to "bone". There is no sense in reporting "bone", so we can access the non-redundant nodes in common via <A HREF="../../../../org/obd/model/stats/SimilarityPair.html#getNonRedundantNodesInCommon()"><CODE>getNonRedundantNodesInCommon()</CODE></A>.
 See also <A HREF="../../../../org/obd/model/stats/SimilarityPair.html#getNonRedundantNodesInSet1()"><CODE>getNonRedundantNodesInSet1()</CODE></A> and <A HREF="../../../../org/obd/model/stats/SimilarityPair.html#getNonRedundantNodesInSet2()"><CODE>getNonRedundantNodesInSet2()</CODE></A>
 
 <h2>Scoring</h2>
 
 With these values calculated, we can assign similarity scores. One measure of similarity is simply the number of nodes in common. However,
 this would mean that genes with high numbers of annotations would be counted as similar by chance. We also want to penalise differences.
 <p>
 The basicSimilarityScore (aka class overlap) is the ratio of nodesInCommon to nodesInUnion
 <pre>
 | nodesInSet1 &cap; nodesInSet2 | / | nodesInSet1 &cup; nodesInSet2 |
 </pre>
 also written as:
 
 <img src="http://www.pubmedcentral.nih.gov/picrender.fcgi?artid=2238903&blobname=gkm806um7.jpg" alt="formula for UI(p,q)"> 
 <p>
 
 See <A HREF="../../../../org/obd/model/stats/SimilarityPair.html#getBasicSimilarityScore()"><CODE>getBasicSimilarityScore()</CODE></A>
 <p>
 Recall that this includes inferred annotations. This is desirable for two reasons: it allows approximate matching for non-exact classes, and it penalises general matches
 in favour of specific matches.
 <p>
 To see why this is true, consider two genes annotated to very specific classes - e.g. geneA to "permeability of ER membrane of pyramidal neuron" vs geneB to "permability
 of ER membrane of interneuron". Although not an exact match, they will have very many inferred classes in common, because they are so deep in the graph. Thus |nodesInCommon|
 will be high
 <p>
 Contrast with distant annotations: these may trivially have nodesInCommon inferred near the top of the ontology graph: {"quality", "organism", "organ"} etc. As these
 nodes have fewer ancestors, |nodesInCommon| will be lower.
 <p>
 This works both ways: differences between the genes may be highly specific too, resulting in a bigger penalty. This may be more pronounced across species, where
 species specific anatomy ontologies are used.
 <p>
 Some studies have found that class overlap metrics are sufficient; however, the graph may be a poor proxy for what constitutes an important match. For that we have to use
 information content
 
 <h3>Information Content Metrics</h3>
 
 The information content of a class is a measure of how "surprised" we are to see it in an annotation.
 <p>
 "organism" has a low IC because it occurs so frequently (taking inference into account). If we see two genes A and B annotated to this it may well be by chance, the
 genes are not necessarily related.
 <p>
 "permeability of ER membrane in pyramidal neuron on CA4 hippocampal region" in contrast has high IC. Two genes A and B annotated to this are less likely to occur by chance
 (although there is of course the possibility of annotation bias: if this was a heavily studied region we would expect to see more)
 <p>
 The IC of a term/class t is:
 <pre>
    IC(t) = -log<SUB>2</SUB>p(t)
 </pre>
 Here p(t) is the probability of an annotated entity chosen at random being annotated (directly or via inference) with t
 <pre>
   p(t) = | annot(t) | / total_annotated_entities
 </pre>
 Use <A HREF="../../../../org/obd/model/stats/SimilarityPair.html#getInformationContent(java.lang.String)"><CODE>getInformationContent(String)</CODE></A> to get the IC for any class in the similarity set.
 <p>
 We can use the IC of a class to improve comparison metrics. One common technique is to take the single most informative (highest IC) class for any given pair of nodes
 in an ontology. For this we use <A HREF="../../../../org/obd/model/stats/SimilarityPair.html#getMaximumInformationContentForNodesInCommon()"><CODE>getMaximumInformationContentForNodesInCommon()</CODE></A>.
 -- However, this masks the kind of similarity where the two nodes share multiple different things in common.
 <p>
 Another metric is to sum IC the non-redundant nodesInCommon. See <A HREF="../../../../org/obd/model/stats/SimilarityPair.html#getInformationContentSumForNRNodesInCommon()"><CODE>getInformationContentSumForNRNodesInCommon()</CODE></A>. We use non-redundant nodes to better
 reflect independent annotations.
 <p>
 The most sophisticated measure here is to get the ratio of IC for the nodesInCommon vs the IC for nodesInUnion. This penalises
 nodes that have differing annotations. We can write this as:
 <img src="http://www.pubmedcentral.nih.gov/picrender.fcgi?artid=2238903&blobname=gkm806um8.jpg" alt="formula for simGIC"/> 
 See <A HREF="../../../../org/obd/model/stats/SimilarityPair.html#getInformationContentRatio()"><CODE>getInformationContentRatio()</CODE></A>
 
 <h2>Inferred annotations and use of the reasoner</h2>
 
 This object makes use of pre-computed reasoner results. It relies on the object which builds the SimilarityPair object to populate a closure map, which indicates the
 ancestors for any given class. See <A HREF="../../../../org/obd/query/AnalysisCapableRepository.html#compareAnnotationsByAnnotatedEntityPair(java.lang.String, java.lang.String)"><CODE>AnalysisCapableRepository.compareAnnotationsByAnnotatedEntityPair(String, String)</CODE></A>.
 <p>
 The most common case is to populate the closure map using the full set of reasoner results, but ignoring the final inferred relation. Thus "cell" is treated as an ancestor of
 both "nucleus" and "T-cell". However, the closure map can also be built to only reflect is_a or partonomic relations, if preferred.
 <p>
 The pre-reasoned results are essential for finding nodesInCommon - annotations do not necessarily match exactly - they may match further
 up the graph. They are also used when calculating nonRedundantNodesInCommmon (<A HREF="../../../../org/obd/model/stats/SimilarityPair.html#getNonRedundantNodesInCommon()"><CODE>getNonRedundantNodesInCommon()</CODE></A>, so we do not report or double-count
 nodes that subsume existing nodes. For example, if nodesInCommon = {neuron, nervous system, morphology-of-tibia, morphology-of-bone} then nonRedundantNodesInCommon =
 {neuron, morphology-of-tibia} because two of the nodes are redundant (assumes the closureMap treates all relations equally. If we build a closure map from is_a only, then
 nervous system is NOT redundant with neuron).
 <p>
 <CODE>org.obd.model.CompositionalDescriptions</CODE> are treated much like other classes. For example, the EQ description "shortened dendrite" is treated as an ontology-less class. It is an
 is_a descendant of "size of cell projection" and of "size". It is a descendant via inheres_in to "dendrite" and "cell projection". It is a descendant via inheres_in_part_of
 to "cell", "organism" etc. The distinctions between the relations are generally lost when the SimilarityPair object is constructed.
 
 <h3>Important point</h3>
 
 The reasoner will only know about ontology classes or pre-existing CompositionalDescriptions. For example, if we have annotations to both
 curvature-of-mouse-tibia and shape-of-human-tibia we would like to use morphology-of-uberon-tibia as the node in common.
 <p>
 However, if the composition morphology-of-uberon-tibia has not been added to the database in advance, then nothing is known about this composition! the reasoner
 does NOT compare all <i>potential</i> compositions as this would be too expensive. The most likely nodeInCommon in this particular case may turn out to be a pre-coordinated
 ontology class such as uberon-tibia.
 <p>
 This is partly a problem as far as compositional descriptions are concerned.
 <p>
 One way around this is to pre-populate the database with likely compositions. The <CODE>PhenotypeHelper</CODE> class does this for phenotype-specific combinations. For
 example, the database is pre-populated with uberon-based compositions whenever there is a species-centric anatomy ontology composition. The reasoner can then place
 this in the subsumption hierarchy correctly, and potentially use this as a nodeInCommon in a SimilarityPair comparison.
 <p>
 One possible future expansion is to make SimilarityPair smarter with respect to <i>potential</i> combinations such that these do not have to be pre-populated.
 
 
 
 <h2>References</h2>
 The following references may be of use:
 <ul>
 <li>
 <a href="http://www.pubmedcentral.nih.gov/articlerender.fcgi?tool=pubmed&pubmedid=17932054">FunSimMat: a comprehensive functional similarity database</a>
 </li>
 <li>
 <a href="http://www.biomedcentral.com/1471-2105/9/327">Gene Ontology term overlap as a measure of gene functional similarity</a>
 </li>
 </ul>
<P>

<P>
<DL>
<DT><B>Author:</B></DT>
  <DD>cjm</DD>
<DT><B>See Also:</B><DD><A HREF="../../../../org/obd/query/AnalysisCapableRepository.html#compareAnnotationsByAnnotatedEntityPair(java.lang.String, java.lang.String)"><CODE>AnalysisCapableRepository.compareAnnotationsByAnnotatedEntityPair(String, String)</CODE></A>, 
<A HREF="../../../../org/obd/query/AnalysisCapableRepository.html#compareAnnotationsBySourcePair(java.lang.String, java.lang.String, java.lang.String)"><CODE>AnalysisCapableRepository.compareAnnotationsBySourcePair(String, String, String)</CODE></A></DL>
<HR>

<P>

<!-- ======== CONSTRUCTOR SUMMARY ======== -->

<A NAME="constructor_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Constructor Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../org/obd/model/stats/SimilarityPair.html#SimilarityPair()">SimilarityPair</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<!-- ========== METHOD SUMMARY =========== -->

<A NAME="method_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Method Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/obd/model/stats/SimilarityPair.html#calculateInverseClosureMap()">calculateInverseClosureMap</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;double</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/obd/model/stats/SimilarityPair.html#getBasicSimilarityScore()">getBasicSimilarityScore</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The basic similarity score is the number of nodes in common divided by the number of nodes in the union.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.Map&lt;java.lang.String,java.util.Set&lt;java.lang.String&gt;&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/obd/model/stats/SimilarityPair.html#getClosureMap()">getClosureMap</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/obd/model/Graph.html" title="class in org.obd.model">Graph</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/obd/model/stats/SimilarityPair.html#getGraph()">getGraph</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/obd/model/stats/SimilarityPair.html#getId1()">getId1</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/obd/model/stats/SimilarityPair.html#getId2()">getId2</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.Double</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/obd/model/stats/SimilarityPair.html#getInformationContent(java.lang.String)">getInformationContent</A></B>(java.lang.String&nbsp;nid)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Given a node identifier, return the information content for that node.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.Double</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/obd/model/stats/SimilarityPair.html#getInformationContentRatio()">getInformationContentRatio</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;information content of intersection set divided by information content of union set</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.Double</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/obd/model/stats/SimilarityPair.html#getInformationContentSumForNRNodesInCommon()">getInformationContentSumForNRNodesInCommon</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&Sigma;<SUB>t &isin; nrNodesInCommon</SUB> IC(t)</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.Map&lt;java.lang.String,java.util.Set&lt;java.lang.String&gt;&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/obd/model/stats/SimilarityPair.html#getInverseClosureMap()">getInverseClosureMap</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.Double</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/obd/model/stats/SimilarityPair.html#getMaximumInformationContentForNodesInCommon()">getMaximumInformationContentForNodesInCommon</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;the maximum shannon information of all nodes in intersection set.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.Map&lt;java.lang.String,java.lang.Double&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/obd/model/stats/SimilarityPair.html#getNodeInformationContentMap()">getNodeInformationContentMap</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.Set&lt;java.lang.String&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/obd/model/stats/SimilarityPair.html#getNodesInCommon()">getNodesInCommon</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;nodesInSet1 &cap; nodesInSet2</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.Set&lt;java.lang.String&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/obd/model/stats/SimilarityPair.html#getNodesInSet1()">getNodesInSet1</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.Set&lt;java.lang.String&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/obd/model/stats/SimilarityPair.html#getNodesInSet2()">getNodesInSet2</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.Set&lt;java.lang.String&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/obd/model/stats/SimilarityPair.html#getNodesInUnion()">getNodesInUnion</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;nodesInSet1 &cup; nodesInSet2</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/obd/model/stats/SimilarityPair.html#getNodeWithMaximumInformationContent()">getNodeWithMaximumInformationContent</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.Set&lt;java.lang.String&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/obd/model/stats/SimilarityPair.html#getNonRedundantNodesInCommon()">getNonRedundantNodesInCommon</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.Set&lt;java.lang.String&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/obd/model/stats/SimilarityPair.html#getNonRedundantNodesInSet1()">getNonRedundantNodesInSet1</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.Set&lt;java.lang.String&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/obd/model/stats/SimilarityPair.html#getNonRedundantNodesInSet2()">getNonRedundantNodesInSet2</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.Set&lt;java.lang.String&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/obd/model/stats/SimilarityPair.html#getPostPartitionedNonRedundantNodesInUnion()">getPostPartitionedNonRedundantNodesInUnion</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Note: this is different from the nonredundant set in the union.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.List&lt;<A HREF="../../../../org/obd/model/stats/ScoredNode.html" title="class in org.obd.model.stats">ScoredNode</A>&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/obd/model/stats/SimilarityPair.html#getScoredNodes()">getScoredNodes</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/obd/model/stats/SimilarityPair.html#getTotalNodesInCommon()">getTotalNodesInCommon</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;| nodesInSet1 &cap; nodesInSet2 |</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/obd/model/stats/SimilarityPair.html#getTotalNodesInSet1()">getTotalNodesInSet1</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/obd/model/stats/SimilarityPair.html#getTotalNodesInSet2()">getTotalNodesInSet2</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/obd/model/stats/SimilarityPair.html#getTotalNodesInUnion()">getTotalNodesInUnion</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;| nodesInSet1 &cup; nodesInSet2 |</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/obd/model/stats/SimilarityPair.html#setClosureMap(java.util.Map)">setClosureMap</A></B>(java.util.Map&lt;java.lang.String,java.util.Set&lt;java.lang.String&gt;&gt;&nbsp;closureMap)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/obd/model/stats/SimilarityPair.html#setFromNodeSetPair(java.util.Collection, java.util.Collection)">setFromNodeSetPair</A></B>(java.util.Collection&lt;java.lang.String&gt;&nbsp;ids1,
                   java.util.Collection&lt;java.lang.String&gt;&nbsp;ids2)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Inititialize based on two sets of class nodes; calculate the set of nodes in common'</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/obd/model/stats/SimilarityPair.html#setGraph(org.obd.model.Graph)">setGraph</A></B>(<A HREF="../../../../org/obd/model/Graph.html" title="class in org.obd.model">Graph</A>&nbsp;graph)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/obd/model/stats/SimilarityPair.html#setId1(java.lang.String)">setId1</A></B>(java.lang.String&nbsp;baseNode)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/obd/model/stats/SimilarityPair.html#setId2(java.lang.String)">setId2</A></B>(java.lang.String&nbsp;targetNode)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/obd/model/stats/SimilarityPair.html#setInformationContent(java.lang.String, java.lang.Double)">setInformationContent</A></B>(java.lang.String&nbsp;nid,
                      java.lang.Double&nbsp;ic)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/obd/model/stats/SimilarityPair.html#setInformationContentRatio(java.lang.Double)">setInformationContentRatio</A></B>(java.lang.Double&nbsp;similarityByInformationContentRatio)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/obd/model/stats/SimilarityPair.html#setInformationContentSumForNRNodesInCommon(java.lang.Double)">setInformationContentSumForNRNodesInCommon</A></B>(java.lang.Double&nbsp;similarityByInformationContentInCommonSum)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/obd/model/stats/SimilarityPair.html#setMaximimumInformationContentForNodesInCommon(double)">setMaximimumInformationContentForNodesInCommon</A></B>(double&nbsp;maximimumInformationContent)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/obd/model/stats/SimilarityPair.html#setNodeInformationContentMap(java.util.Map)">setNodeInformationContentMap</A></B>(java.util.Map&lt;java.lang.String,java.lang.Double&gt;&nbsp;nodeInformationContentMap)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/obd/model/stats/SimilarityPair.html#setNodesInCommon(java.util.Set)">setNodesInCommon</A></B>(java.util.Set&lt;java.lang.String&gt;&nbsp;nodesInCommon)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/obd/model/stats/SimilarityPair.html#setNodesInSet1(java.util.Set)">setNodesInSet1</A></B>(java.util.Set&lt;java.lang.String&gt;&nbsp;nodesInSet1)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/obd/model/stats/SimilarityPair.html#setNodesInSet2(java.util.Set)">setNodesInSet2</A></B>(java.util.Set&lt;java.lang.String&gt;&nbsp;nodesInSet2)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/obd/model/stats/SimilarityPair.html#setNodesInUnion(java.util.Set)">setNodesInUnion</A></B>(java.util.Set&lt;java.lang.String&gt;&nbsp;nodesInUnion)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/obd/model/stats/SimilarityPair.html#setNodeWithMaximumInformationContent(java.lang.String)">setNodeWithMaximumInformationContent</A></B>(java.lang.String&nbsp;nodeWithMaximimumInformationContent)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/obd/model/stats/SimilarityPair.html#setScoredNodes(java.util.List)">setScoredNodes</A></B>(java.util.List&lt;<A HREF="../../../../org/obd/model/stats/ScoredNode.html" title="class in org.obd.model.stats">ScoredNode</A>&gt;&nbsp;scoredNodes)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/obd/model/stats/SimilarityPair.html#setTotalNodesInCommon(int)">setTotalNodesInCommon</A></B>(int&nbsp;totalNodesInCommon)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/obd/model/stats/SimilarityPair.html#setTotalNodesInSet1(int)">setTotalNodesInSet1</A></B>(int&nbsp;totalNodesInSet1)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/obd/model/stats/SimilarityPair.html#setTotalNodesInSet2(int)">setTotalNodesInSet2</A></B>(int&nbsp;totalNodesInSet2)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/obd/model/stats/SimilarityPair.html#setTotalNodesInUnion(int)">setTotalNodesInUnion</A></B>(int&nbsp;totalNodes)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/obd/model/stats/SimilarityPair.html#sortScoredNodes()">sortScoredNodes</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/obd/model/stats/SimilarityPair.html#toString()">toString</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_java.lang.Object"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Methods inherited from class java.lang.Object</B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE>clone, equals, finalize, getClass, hashCode, notify, notifyAll, wait, wait, wait</CODE></TD>
</TR>
</TABLE>
&nbsp;
<P>

<!-- ========= CONSTRUCTOR DETAIL ======== -->

<A NAME="constructor_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Constructor Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="SimilarityPair()"><!-- --></A><H3>
SimilarityPair</H3>
<PRE>
public <B>SimilarityPair</B>()</PRE>
<DL>
</DL>

<!-- ============ METHOD DETAIL ========== -->

<A NAME="method_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Method Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="getGraph()"><!-- --></A><H3>
getGraph</H3>
<PRE>
public <A HREF="../../../../org/obd/model/Graph.html" title="class in org.obd.model">Graph</A> <B>getGraph</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="setGraph(org.obd.model.Graph)"><!-- --></A><H3>
setGraph</H3>
<PRE>
public void <B>setGraph</B>(<A HREF="../../../../org/obd/model/Graph.html" title="class in org.obd.model">Graph</A>&nbsp;graph)</PRE>
<DL>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="getId1()"><!-- --></A><H3>
getId1</H3>
<PRE>
public java.lang.String <B>getId1</B>()</PRE>
<DL>
<DD><DL>

<DT><B>Returns:</B><DD>identifier of first node in pair (may be ontology class, annotation source, annotated entity, ...)
 This node is associated with a set of other nodes (e.g. annotated entities)</DL>
</DD>
</DL>
<HR>

<A NAME="setId1(java.lang.String)"><!-- --></A><H3>
setId1</H3>
<PRE>
public void <B>setId1</B>(java.lang.String&nbsp;baseNode)</PRE>
<DL>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="getId2()"><!-- --></A><H3>
getId2</H3>
<PRE>
public java.lang.String <B>getId2</B>()</PRE>
<DL>
<DD><DL>

<DT><B>Returns:</B><DD>identifier of second node in pair (may be ontology class, annotation source, annotated entity, ...)</DL>
</DD>
</DL>
<HR>

<A NAME="setId2(java.lang.String)"><!-- --></A><H3>
setId2</H3>
<PRE>
public void <B>setId2</B>(java.lang.String&nbsp;targetNode)</PRE>
<DL>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="getTotalNodesInUnion()"><!-- --></A><H3>
getTotalNodesInUnion</H3>
<PRE>
public int <B>getTotalNodesInUnion</B>()</PRE>
<DL>
<DD>| nodesInSet1 &cup; nodesInSet2 |
<P>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="setTotalNodesInUnion(int)"><!-- --></A><H3>
setTotalNodesInUnion</H3>
<PRE>
public void <B>setTotalNodesInUnion</B>(int&nbsp;totalNodes)</PRE>
<DL>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="getTotalNodesInCommon()"><!-- --></A><H3>
getTotalNodesInCommon</H3>
<PRE>
public int <B>getTotalNodesInCommon</B>()</PRE>
<DL>
<DD>| nodesInSet1 &cap; nodesInSet2 |
<P>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="setTotalNodesInCommon(int)"><!-- --></A><H3>
setTotalNodesInCommon</H3>
<PRE>
public void <B>setTotalNodesInCommon</B>(int&nbsp;totalNodesInCommon)</PRE>
<DL>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="getTotalNodesInSet1()"><!-- --></A><H3>
getTotalNodesInSet1</H3>
<PRE>
public int <B>getTotalNodesInSet1</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="setTotalNodesInSet1(int)"><!-- --></A><H3>
setTotalNodesInSet1</H3>
<PRE>
public void <B>setTotalNodesInSet1</B>(int&nbsp;totalNodesInSet1)</PRE>
<DL>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="getTotalNodesInSet2()"><!-- --></A><H3>
getTotalNodesInSet2</H3>
<PRE>
public int <B>getTotalNodesInSet2</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="setTotalNodesInSet2(int)"><!-- --></A><H3>
setTotalNodesInSet2</H3>
<PRE>
public void <B>setTotalNodesInSet2</B>(int&nbsp;totalNodesInSet2)</PRE>
<DL>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="getNodesInCommon()"><!-- --></A><H3>
getNodesInCommon</H3>
<PRE>
public java.util.Set&lt;java.lang.String&gt; <B>getNodesInCommon</B>()</PRE>
<DL>
<DD>nodesInSet1 &cap; nodesInSet2
<P>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="getNonRedundantNodesInCommon()"><!-- --></A><H3>
getNonRedundantNodesInCommon</H3>
<PRE>
public java.util.Set&lt;java.lang.String&gt; <B>getNonRedundantNodesInCommon</B>()</PRE>
<DL>
<DD><DL>

<DT><B>Returns:</B><DD>all n : n &isin; nodesInSet1 &cap; nodesInSet2 & invclosure(n) &notin; nodesInSet1 &cap; nodesInSet2</DL>
</DD>
</DL>
<HR>

<A NAME="setNodesInCommon(java.util.Set)"><!-- --></A><H3>
setNodesInCommon</H3>
<PRE>
public void <B>setNodesInCommon</B>(java.util.Set&lt;java.lang.String&gt;&nbsp;nodesInCommon)</PRE>
<DL>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="getNodesInSet1()"><!-- --></A><H3>
getNodesInSet1</H3>
<PRE>
public java.util.Set&lt;java.lang.String&gt; <B>getNodesInSet1</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="getNonRedundantNodesInSet1()"><!-- --></A><H3>
getNonRedundantNodesInSet1</H3>
<PRE>
public java.util.Set&lt;java.lang.String&gt; <B>getNonRedundantNodesInSet1</B>()</PRE>
<DL>
<DD><DL>

<DT><B>Returns:</B><DD>all n : n &isin; nodesInSet1  & invclosure(n) &notin; nodesInSet1</DL>
</DD>
</DL>
<HR>

<A NAME="setNodesInSet1(java.util.Set)"><!-- --></A><H3>
setNodesInSet1</H3>
<PRE>
public void <B>setNodesInSet1</B>(java.util.Set&lt;java.lang.String&gt;&nbsp;nodesInSet1)</PRE>
<DL>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="getNodesInSet2()"><!-- --></A><H3>
getNodesInSet2</H3>
<PRE>
public java.util.Set&lt;java.lang.String&gt; <B>getNodesInSet2</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="getNonRedundantNodesInSet2()"><!-- --></A><H3>
getNonRedundantNodesInSet2</H3>
<PRE>
public java.util.Set&lt;java.lang.String&gt; <B>getNonRedundantNodesInSet2</B>()</PRE>
<DL>
<DD><DL>

<DT><B>Returns:</B><DD>all n : n &isin; nodesInSet2  & invclosure(n) &notin; nodesInSet2</DL>
</DD>
</DL>
<HR>

<A NAME="setNodesInSet2(java.util.Set)"><!-- --></A><H3>
setNodesInSet2</H3>
<PRE>
public void <B>setNodesInSet2</B>(java.util.Set&lt;java.lang.String&gt;&nbsp;nodesInSet2)</PRE>
<DL>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="getNodesInUnion()"><!-- --></A><H3>
getNodesInUnion</H3>
<PRE>
public java.util.Set&lt;java.lang.String&gt; <B>getNodesInUnion</B>()</PRE>
<DL>
<DD>nodesInSet1 &cup; nodesInSet2
<P>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="getPostPartitionedNonRedundantNodesInUnion()"><!-- --></A><H3>
getPostPartitionedNonRedundantNodesInUnion</H3>
<PRE>
public java.util.Set&lt;java.lang.String&gt; <B>getPostPartitionedNonRedundantNodesInUnion</B>()</PRE>
<DL>
<DD>Note: this is different from the nonredundant set in the union. Instead
 we take 3 sets: NRnodesInSetBoth, NRnodesInSet1, NRnodesInSet2
 (these are all guaranteed internally non-redundant) and take the union of these
 sets. The final set will NOT be itself non-redundant.
 For example, NRset1 may contain "left hand", NRset2 may contain "right hand" and
 NRinBoth may contain "hand". The post-partitioned union will contain all 3,
 despite the fact they are not non-redundant with eachother, only within their
 respective sets
<P>
<DD><DL>

<DT><B>Returns:</B><DD>NRnodesInSetBoth &cup; NRnodesInSet1 &cup; NRnodesInSet2 &cup;</DL>
</DD>
</DL>
<HR>

<A NAME="setNodesInUnion(java.util.Set)"><!-- --></A><H3>
setNodesInUnion</H3>
<PRE>
public void <B>setNodesInUnion</B>(java.util.Set&lt;java.lang.String&gt;&nbsp;nodesInUnion)</PRE>
<DL>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="getClosureMap()"><!-- --></A><H3>
getClosureMap</H3>
<PRE>
public java.util.Map&lt;java.lang.String,java.util.Set&lt;java.lang.String&gt;&gt; <B>getClosureMap</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="setClosureMap(java.util.Map)"><!-- --></A><H3>
setClosureMap</H3>
<PRE>
public void <B>setClosureMap</B>(java.util.Map&lt;java.lang.String,java.util.Set&lt;java.lang.String&gt;&gt;&nbsp;closureMap)</PRE>
<DL>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="getInverseClosureMap()"><!-- --></A><H3>
getInverseClosureMap</H3>
<PRE>
public java.util.Map&lt;java.lang.String,java.util.Set&lt;java.lang.String&gt;&gt; <B>getInverseClosureMap</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="calculateInverseClosureMap()"><!-- --></A><H3>
calculateInverseClosureMap</H3>
<PRE>
public void <B>calculateInverseClosureMap</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="setInformationContent(java.lang.String, java.lang.Double)"><!-- --></A><H3>
setInformationContent</H3>
<PRE>
public void <B>setInformationContent</B>(java.lang.String&nbsp;nid,
                                  java.lang.Double&nbsp;ic)</PRE>
<DL>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="getInformationContent(java.lang.String)"><!-- --></A><H3>
getInformationContent</H3>
<PRE>
public java.lang.Double <B>getInformationContent</B>(java.lang.String&nbsp;nid)</PRE>
<DL>
<DD>Given a node identifier, return the information content for that node. Must be a node within the set being analyzed (or the closure of that set).
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>nid</CODE> - 
<DT><B>Returns:</B><DD>-log2(p(nid))</DL>
</DD>
</DL>
<HR>

<A NAME="getNodeInformationContentMap()"><!-- --></A><H3>
getNodeInformationContentMap</H3>
<PRE>
public java.util.Map&lt;java.lang.String,java.lang.Double&gt; <B>getNodeInformationContentMap</B>()</PRE>
<DL>
<DD><DL>

<DT><B>Returns:</B><DD>map between node IDs and information content<DT><B>See Also:</B><DD><A HREF="../../../../org/obd/model/stats/SimilarityPair.html#getInformationContent(java.lang.String)"><CODE>getInformationContent(String)</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="setNodeInformationContentMap(java.util.Map)"><!-- --></A><H3>
setNodeInformationContentMap</H3>
<PRE>
public void <B>setNodeInformationContentMap</B>(java.util.Map&lt;java.lang.String,java.lang.Double&gt;&nbsp;nodeInformationContentMap)</PRE>
<DL>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="sortScoredNodes()"><!-- --></A><H3>
sortScoredNodes</H3>
<PRE>
public void <B>sortScoredNodes</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="getScoredNodes()"><!-- --></A><H3>
getScoredNodes</H3>
<PRE>
public java.util.List&lt;<A HREF="../../../../org/obd/model/stats/ScoredNode.html" title="class in org.obd.model.stats">ScoredNode</A>&gt; <B>getScoredNodes</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="setScoredNodes(java.util.List)"><!-- --></A><H3>
setScoredNodes</H3>
<PRE>
public void <B>setScoredNodes</B>(java.util.List&lt;<A HREF="../../../../org/obd/model/stats/ScoredNode.html" title="class in org.obd.model.stats">ScoredNode</A>&gt;&nbsp;scoredNodes)</PRE>
<DL>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="getMaximumInformationContentForNodesInCommon()"><!-- --></A><H3>
getMaximumInformationContentForNodesInCommon</H3>
<PRE>
public java.lang.Double <B>getMaximumInformationContentForNodesInCommon</B>()</PRE>
<DL>
<DD>the maximum shannon information of all nodes in intersection set.
 <p>
 note that this object can not calculate this metric by itself - it must be populated externally
<P>
<DD><DL>

<DT><B>Returns:</B><DD>max(IC(t) forall t in both)</DL>
</DD>
</DL>
<HR>

<A NAME="setMaximimumInformationContentForNodesInCommon(double)"><!-- --></A><H3>
setMaximimumInformationContentForNodesInCommon</H3>
<PRE>
public void <B>setMaximimumInformationContentForNodesInCommon</B>(double&nbsp;maximimumInformationContent)</PRE>
<DL>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="getNodeWithMaximumInformationContent()"><!-- --></A><H3>
getNodeWithMaximumInformationContent</H3>
<PRE>
public java.lang.String <B>getNodeWithMaximumInformationContent</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="setNodeWithMaximumInformationContent(java.lang.String)"><!-- --></A><H3>
setNodeWithMaximumInformationContent</H3>
<PRE>
public void <B>setNodeWithMaximumInformationContent</B>(java.lang.String&nbsp;nodeWithMaximimumInformationContent)</PRE>
<DL>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="setInformationContentRatio(java.lang.Double)"><!-- --></A><H3>
setInformationContentRatio</H3>
<PRE>
public void <B>setInformationContentRatio</B>(java.lang.Double&nbsp;similarityByInformationContentRatio)</PRE>
<DL>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="getInformationContentRatio()"><!-- --></A><H3>
getInformationContentRatio</H3>
<PRE>
public java.lang.Double <B>getInformationContentRatio</B>()</PRE>
<DL>
<DD>information content of intersection set divided by information content of union set
 <p>
 See:
 <pre>
 Pesquita, C.;Faria, D.;Bastos, H.;Falco, AO.; Couto, FM. 
 <i>Evaluating GO-based semantic similarity measures. Proceedings of the 10th Annual Bio-Ontologies Meeting (Bio-Ontologies 2007). </i>
 Stevens R, Lord P, McEntire R, Sansone S-A. , editors. Austria: Vienna; 2007. pp. 3740.
 </pre>
 <p>
 Original Formula:
 <img src="http://www.pubmedcentral.nih.gov/picrender.fcgi?artid=2238903&blobname=gkm806um8.jpg" alt="formula for simGIC"/>
 <p>
 from: <a href="http://www.pubmedcentral.nih.gov/articlerender.fcgi?tool=pubmed&pubmedid=17932054">FunSimMat: a comprehensive functional similarity database</a>
 <p>
 <b>NOTE:</b> we only include non-redundant nodes in the calculation
 <p>
 note that this object can not calculate this metric by itself - it must be populated externally
<P>
<DD><DL>

<DT><B>Returns:</B><DD>sum(IC(t) where t in NRnodesInCommon) / sum(IC(t) where t in NRnodesInUnion)<DT><B>See Also:</B><DD><A HREF="../../../../org/obd/query/AnalysisCapableRepository.html#calculateInformationContentMetrics(org.obd.model.stats.SimilarityPair)"><CODE>AnalysisCapableRepository.calculateInformationContentMetrics(SimilarityPair)</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="getInformationContentSumForNRNodesInCommon()"><!-- --></A><H3>
getInformationContentSumForNRNodesInCommon</H3>
<PRE>
public java.lang.Double <B>getInformationContentSumForNRNodesInCommon</B>()</PRE>
<DL>
<DD>&Sigma;<SUB>t &isin; nrNodesInCommon</SUB> IC(t)
<P>
<DD><DL>

<DT><B>Returns:</B><DD>sum( IC : forall IC in NonRedundantNodesInCommon )</DL>
</DD>
</DL>
<HR>

<A NAME="setInformationContentSumForNRNodesInCommon(java.lang.Double)"><!-- --></A><H3>
setInformationContentSumForNRNodesInCommon</H3>
<PRE>
public void <B>setInformationContentSumForNRNodesInCommon</B>(java.lang.Double&nbsp;similarityByInformationContentInCommonSum)</PRE>
<DL>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="setFromNodeSetPair(java.util.Collection, java.util.Collection)"><!-- --></A><H3>
setFromNodeSetPair</H3>
<PRE>
public void <B>setFromNodeSetPair</B>(java.util.Collection&lt;java.lang.String&gt;&nbsp;ids1,
                               java.util.Collection&lt;java.lang.String&gt;&nbsp;ids2)</PRE>
<DL>
<DD>Inititialize based on two sets of class nodes; calculate the set of nodes in common'
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>ids1</CODE> - <DD><CODE>ids2</CODE> - </DL>
</DD>
</DL>
<HR>

<A NAME="getBasicSimilarityScore()"><!-- --></A><H3>
getBasicSimilarityScore</H3>
<PRE>
public double <B>getBasicSimilarityScore</B>()</PRE>
<DL>
<DD>The basic similarity score is the number of nodes in common divided by the number of nodes in the union. We assume that this has been pre-assigned,
 taking into account closure over relations if required.
 Formula: <img src="http://www.pubmedcentral.nih.gov/picrender.fcgi?artid=2238903&blobname=gkm806um7.jpg" alt="formula for UI(p,q)"> in
 <a href="http://www.pubmedcentral.nih.gov/articlerender.fcgi?tool=pubmed&pubmedid=17932054">FunSimMat: a comprehensive functional similarity database</a>
 Note: at this time, this number reflects the complete set - no filtering is done for redundancy.
 This is the desired behavior, as the redundant nodes are in proportion to semantic similarity.
 For example, if a highly specific node is shared in common, then its ancestors are also included, boosting
 the score
<P>
<DD><DL>

<DT><B>Returns:</B><DD>totalNodesInCommon / totalNodesInUnion</DL>
</DD>
</DL>
<HR>

<A NAME="toString()"><!-- --></A><H3>
toString</H3>
<PRE>
public java.lang.String <B>toString</B>()</PRE>
<DL>
<DD><DL>
<DT><B>Overrides:</B><DD><CODE>toString</CODE> in class <CODE>java.lang.Object</CODE></DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<!-- ========= END OF CLASS DATA ========= -->
<HR>


<!-- ======= START OF BOTTOM NAVBAR ====== -->
<A NAME="navbar_bottom"><!-- --></A>
<A HREF="#skip-navbar_bottom" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_bottom_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="class-use/SimilarityPair.html"><FONT CLASS="NavBarFont1"><B>Use</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../../../../org/obd/model/stats/ScoredNodePair.html" title="class in org.obd.model.stats"><B>PREV CLASS</B></A>&nbsp;
&nbsp;<A HREF="../../../../org/obd/model/stats/SummaryStatistics.html" title="class in org.obd.model.stats"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../../index.html?org/obd/model/stats/SimilarityPair.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="SimilarityPair.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;FIELD&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;FIELD&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_bottom"></A>
<!-- ======== END OF BOTTOM NAVBAR ======= -->

<HR>

		            <i>Copyright &#169; 2008 All Rights Reserved.</i>
		        
</BODY>
</HTML>
